app/Exceptions/ValidationExceptionRenderer.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace App\Exceptions;
+
+use App\Services\System\TranslationsService;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Support\Str;
+use Illuminate\Validation\Rules\Password;
+use Illuminate\Validation\ValidationException;
+use ReflectionClass;
+use ReflectionException;
+
+class ValidationExceptionRenderer
+{
+    private array $multiTypeRules;
+    private array $multiTypeTypes;
+
+    public function __construct()
+    {
+        $this->multiTypeRules = array_flip(['between', 'gt', 'gte', 'lt', 'lte', 'max', 'min', 'size']);
+        $this->multiTypeTypes = array_flip(['array', 'file', 'numeric', 'string']);
+    }
+
+    private function matchType(array $fieldRules): string
+    {
+        foreach ($fieldRules as $fieldRule) {
+            if (array_key_exists($fieldRule, $this->multiTypeTypes)) {
+                return $fieldRule;
+            }
+        }
+        return 'string';
+    }
+
+    private function renderOrThrow(ValidationException $originalException, ApiValidationException $exception): void
+    {
+        $validator = $originalException->validator;
+        $validatorReflection = new ReflectionClass($validator);
+        $defaultTranslation = TranslationsService::defaultAppValidationTranslation();
+        try {
+            $rules = $validatorReflection->getProperty('rules')->getValue($validator);
+        } catch (ReflectionException) {
+            throw ConfigExceptionFactory::reflectionRules();
+        }
+        foreach ($validator->failed() as $field => $fieldErrors) {
```

Comment:
może warto podzielić to co się tu dzieje wewnątrz foreach na mniejsze funkcje? troszkę ciężko się to czyta wszystko naraz ;d

Reply:
niech będzie :D